tests/trace/test_dataset.py

```diff
@@ -222,3 +224,388 @@ def test_add_rows_to_unsaved_dataset(client):
+        {"id": 2, "val": 4, "name": "Item 2", "product": 8},
+        {"id": 3, "val": 6, "name": "Item 3", "product": 18},
+        {"id": 4, "val": 8, "name": "Item 4", "product": 32},
+    ]
+    assert list(mapped_ds) == expected_rows
+
+
+def test_dataset_map_with_scalar_return(client):
+    """Test that mapping with a scalar return value raises a TypeError."""
+    original_rows = [{"id": i, "val": i * 2} for i in range(5)]
+    ds = weave.Dataset(rows=original_rows)
+
+    # Function that takes specific columns and returns a scalar
+    def sum_values(id, val):
+        return id + val
+
+    # Should raise TypeError because function returns a scalar, not a dict
+    with pytest.raises(TypeError, match="Function must return a dictionary"):
+        ds.map(sum_values)
+
+    # Correct version that returns a dictionary
+    def sum_values_dict(id, val):
+        return {"sum": id + val}
+
+    # This should work
+    mapped_ds = ds.map(sum_values_dict)
+    assert mapped_ds[2]["sum"] == 6  # 2 + 4
+
+
+def test_dataset_map_with_default_param(client):
+    """Test mapping a function with default parameter values."""
+    original_rows = [{"id": i, "val": i * 2} for i in range(5)]
+    ds = weave.Dataset(rows=original_rows)
+
+    # Function with a default parameter
+    def process_with_default(val, multiplier=10):
+        return {"scaled": val * multiplier}
+
+    mapped_ds = ds.map(process_with_default)
+
+    assert len(mapped_ds) == 5
+    expected_rows = [
+        {"id": 0, "val": 0, "scaled": 0},
+        {"id": 1, "val": 2, "scaled": 20},
+        {"id": 2, "val": 4, "scaled": 40},
+        {"id": 3, "val": 6, "scaled": 60},
+        {"id": 4, "val": 8, "scaled": 80},
+    ]
+    assert list(mapped_ds) == expected_rows
+
+
+def test_dataset_map_missing_param(client):
+    """Test mapping a function that expects a parameter not in the dataset."""
+    original_rows = [{"id": i, "val": i * 2} for i in range(5)]
+    ds = weave.Dataset(rows=original_rows)
+
+    # Function expecting a parameter not in the dataset
+    def needs_missing_param(val, missing_param):
+        return {"result": val + missing_param}
+
+    # Should raise an error because missing_param is not in the dataset
+    with pytest.raises(ValueError, match="Function expects parameter 'missing_param'"):
+        ds.map(needs_missing_param)
+
+
+def test_dataset_map_with_lambda(client):
+    """Test mapping with lambda functions."""
+    ds = weave.Dataset(rows=[{"a": i, "b": i * 2} for i in range(5)])
+
+    # Lambda that takes specific params and returns a dictionary
+    mapped_ds = ds.map(lambda a, b: {"sum": a + b})
+    assert len(mapped_ds) == 5
+    assert mapped_ds[0] == {"a": 0, "b": 0, "sum": 0}
+    assert mapped_ds[2] == {"a": 2, "b": 4, "sum": 6}
+
+    # Lambda with a scalar return value should raise TypeError
+    with pytest.raises(TypeError, match="Function must return a dictionary"):
+        ds.map(lambda a, b: a * b)
```

Comment:
is it too magical to intercept a lambda returning a scalar and instead returning `{"result": a *b}`?

Reply:
it was like that early, but I don't know rick, if they already have key called `results` we would magically overwrite it?